uaclient/system.py

```diff
@@ -188,7 +188,19 @@ def get_virt_type() -> str:
         out, _ = subp(["systemd-detect-virt"])
         return out.strip()
     except exceptions.ProcessExecutionError:
-        return ""
+        # The main known place where that will fail is in a docker/podman
+        # container that doesn't have it installed. So we look for hints
+        # of that situation to report it accurately.
+        try:
+            proc_1_cgroup = load_file("/proc/1/cgroup")
+            if "docker" in proc_1_cgroup or "buildkit" in proc_1_cgroup:
+                return "docker"
+            elif "buildah" in proc_1_cgroup:
+                return "podman"
+            else:
+                return ""
```

Comment:
Those these names appear in specific patterns in the string (i.e. beginning or end of the line) ? If yes, should we have a more restrictive regex for them or do you think name clashing is really unlikely here ?

Reply:
here is what the file looks like in each situation, using the following dockerfile:
```
from ubuntu:focal
run cat /proc/1/cgroup | tee result.txt
```

docker build .
```
13:perf_event:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
12:blkio:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
11:rdma:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
10:pids:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
9:cpuset:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
8:net_cls,net_prio:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
7:devices:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
6:freezer:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
5:cpu,cpuacct:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
4:hugetlb:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
3:memory:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
2:misc:/
1:name=systemd:/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
0::/docker/43820adb8df766d314faece4ddf62332a0eee1fc1c5154ffd4b883fb88cb1232
```

env DOCKER_BUILDKIT=1 docker build . -t test
docker run -it test cat /result.txt
```
13:perf_event:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
12:blkio:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
11:rdma:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
10:pids:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
9:cpuset:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
8:net_cls,net_prio:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
7:devices:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
6:freezer:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
5:cpu,cpuacct:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
4:hugetlb:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
3:memory:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
2:misc:/
1:name=systemd:/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
0::/docker/buildkit/mc0szzrnjv8hx2hqwzvmzriq5
```

podman build .
```
13:perf_event:/
12:blkio:/user.slice
11:rdma:/
10:pids:/user.slice/user-1000.slice/user@1000.service
9:cpuset:/
8:net_cls,net_prio:/
7:devices:/user.slice
6:freezer:/
5:cpu,cpuacct:/user.slice
4:hugetlb:/
3:memory:/user.slice/user-1000.slice/user@1000.service
2:misc:/
1:name=systemd:/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.kde.konsole-1b54bbb6634a43a78eb7033cd3b3ac17.scope/buildah-buildah2868387759
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.kde.konsole-1b54bbb6634a43a78eb7033cd3b3ac17.scope
```

The least obvious one is the podman one where "buildah" only shows up in the "name" field.

So, we could try to make things more strict, but my thought was that it is really unlikely for a clash here. It would only happen in an environment where `systemd-detect-virt` wasn't installed (so we fall back to this function), then the user would have to be doing something tricky/non-standard for one of the magic words to show up there.

And since we're only doing this to get a sense of usage, not for strict tracking, I think its okay.